presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/CachingJdbcClient.java

```diff
@@ -189,12 +185,21 @@ public boolean isLimitGuaranteed(ConnectorSession session)
             return cachedTableHandle;
         }
         Optional<JdbcTableHandle> tableHandle = delegate.getTableHandle(session, schemaTableName);
-        if (tableHandle.isPresent() || cacheMissing) {
+        if ((tableHandle.isPresent() && canBeCached(tableHandle.get())) || (tableHandle.isEmpty() && cacheMissing)) {
             tableHandleCache.put(key, tableHandle);
         }
         return tableHandle;
     }
 
+    private static boolean canBeCached(JdbcTableHandle tableHandle)
+    {
+        // Cannot cache JdbcColumnHandle objects, since their Type may depend on session properties
+        return tableHandle.getGroupingSets().isEmpty() &&
+                tableHandle.getColumns().isEmpty() &&
+                // Synthetic column handle is not expected here, so if one is found it's safer not to cache it
+                !tableHandle.isSynthetic();
```

Comment:
extract `denyCachingSyntheticTableHandle`

Reply:
wdym here?

Reply:
I'd prefer to have method called `itsOkToCacheTableHandleUnlessItContainsColumnHandleObjectsInside` than to rely that much on comments description and code itself. Yes, my example is a little bit exaggerated, but that way it's obvious what's going on underneath. Yes, it's more about `how` than `what` (imperative than declarative) but still you utilize comments for the same sake. 
Please consider my comment from that point of view that I had hard times to understand what's going on here from the first glance, and this your comment https://github.com/prestosql/presto/pull/6161/files#r533357154 helped me a lot.
```
    private static boolean canBeCached(JdbcTableHandle tableHandle)
    {
        return itsOkToCacheUnlessItContainsColumnHandleObjectsInside(tableHandle)
                && denyCachingSynthetic(tableHandle);
    }

    private static boolean itsOkToCacheUnlessItContainsColumnHandleObjectsInside(JdbcTableHandle tableHandle)
    {
        // Cannot cache JdbcColumnHandle objects, since their Type may depend on session properties
        return tableHandle.getGroupingSets().isEmpty() &&
                tableHandle.getColumns().isEmpty();
    }

    private static boolean denyCachingSynthetic(JdbcTableHandle tableHandle)
    {
         // Synthetic column handle is not expected here, so if one is found it's safer not to cache it
        return !tableHandle.isSynthetic();
    }
```